pkg/i2gw/providers/common/utils.go

```diff
@@ -155,3 +155,38 @@ func groupIngressPathsByMatchKey(rules []ingressRule) orderedIngressPathsByMatch
 func PtrTo[T any](a T) *T {
 	return &a
 }
+
+type uniqueBackendRefsKey struct {
+	Name      gatewayv1beta1.ObjectName
+	Namespace gatewayv1beta1.Namespace
+	Port      gatewayv1beta1.PortNumber
+	Group     gatewayv1beta1.Group
+	Kind      gatewayv1beta1.Kind
+}
+
+// DeDuplicateBackendRefs removes duplicate backendRefs from a list of backendRefs.
+func DeDuplicateBackendRefs(backendRefs []gatewayv1beta1.HTTPBackendRef) []gatewayv1beta1.HTTPBackendRef {
+	var deDuplicatedBackendRefs []gatewayv1beta1.HTTPBackendRef
+	uniqueBackendRefs := map[uniqueBackendRefsKey]struct{}{}
+
+	for _, backendRef := range backendRefs {
+		var k uniqueBackendRefsKey
+
+		if backendRef.Port != nil {
+			k.Name = backendRef.Name
+			k.Port = *backendRef.Port
+		} else {
+			k.Name = backendRef.Name
+			k.Group = *backendRef.Group
+			k.Kind = *backendRef.Kind
+		}
```

Comment:
I went with @LiorLieberman 's suggestion as shown here: https://github.com/kubernetes-sigs/ingress2gateway/pull/104#discussion_r1425115085

Reply:
I see, what I suggest is to always enforce group and kind, and check their validity to improve robustness
```suggestion
		group := gatewayv1.Group("")
		kind := gatewayv1.Kind("Service")
		// "" and "core" are equivalent, and here we need consistency
		if backendRef.Group != nil && *backendRef.Group != "core" {
			group = *backendRef.Group
		}
		if backendRef.Kind != nil {
			kind = *backendRef.Kind
		}

		k.Name = backendRef.Name
		k.Group = group
		k.Kind = kind
		if backendRef.Port != nil {
			k.Port = *backendRef.Port
		}
```

Reply:
@mlavacca the suggestion was aligned with the way construct BackendRefs in [toBackendRef](https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/common/converter.go#L371) function.

I think you have a valid point but we need to keep this too consistent. 

Also for the `group` logic you suggested. cant we simply do this ?

```go
if backendRef.Group != nil  {
	group = *backendRef.Group
} else {
        group = gatewayv1.Group("")
}
```

Reply:
@mlavacca @LiorLieberman, what should we do here? Should we keep it as it is?

Reply:
> @mlavacca the suggestion was aligned with the way construct BackendRefs in [toBackendRef](https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/common/converter.go#L371) function.
> 
> I think you have a valid point but we need to keep this too consistent.
> 

I see, I agree we should be consistent. @pottekkat do you mind applying the same logic I suggested to the place @LiorLieberman pointed out?

> Also for the `group` logic you suggested. cant we simply do this ?
> 
> ```go
> if backendRef.Group != nil  {
> 	group = *backendRef.Group
> } else {
>         group = gatewayv1.Group("")
> }
> ```

We can't simply do this because "core" and "" are semantically the same, but if we mix them we end up having duplicated backendrefs. So, we need consistency and force "core" to be "" or "" to be "core", and this is just a matter of choice.